### DISCUSSION TITLE
SQLx from a function?

### DISCUSSION THREAD

**Discussion by ElhamAryanpur** [mlua-rs/mlua#440] [Q&A] [answered]
Hey there! I wanted to call to an SQLx database from within an mlua async function.

Biggest issue I came across was that I think mlua uses futures-utils and tokio based sqlx straight up refused to work, and async-std based sqlx freezes the compiler, and doesn't return anything.

Is there any way to run a tokio or async-std based library in mlua async functions?

A large amount of my codebase depends on lua and this is a hard need...

**Comment by ElhamAryanpur** [comment, marked as answer]
Solution for those seeking is to use Tokio's handle. Possible through a static `once_cell`. And then spawn an async thread to do the task you want.

this stackoverflow thread can be helpful: https://stackoverflow.com/questions/61292425/how-to-run-an-asynchronous-task-from-a-non-main-thread-in-tokio

**Comment by khvzak** [comment]
Do you use module mode?
Otherwise unlikely you need to do anything else apart from having 
```rust
#[tokio::main]
async fn main() {}
```

**Reply by ElhamAryanpur** [reply]
nope. I had a function like this:

```rs
            lua.create_async_function(
                move |lua, (service_name, data): (String, String)| async {
                    let result = call_service::call_service(lua, service_name, data).await;
                    println!("called service result: {result:#?}");

                    result
                },
            )
```

where the result would be a sqlx query result.

Spawning a tokio task with the handle seems to do the trick, except now it seems like joining the task to get a result is not awaiting at all!
```rs
async fn join_result<T>(join_handle: tokio::task::JoinHandle<T>) -> Option<T> {
    let mut final_result: Option<T> = None;

    loop {
        // wait till the task is finished
        if join_handle.is_finished() {
            // get the results if its without error
            if let Ok(result) = join_handle.await {
                final_result = Some(result);
            }

            break;
        }
    }

    final_result
}

// .. then used as

if let Some(result) = join_result::<MyType>(result).await {
    lua.to_value(&result)
} else {
    Ok(mlua::Value::Nil)
}
```

@khvzak do you think the task not awaiting may be mlua bug? I think its by most chance tokio's side.

**Reply by cheesycod** [reply]
@ElhamAryanpur Which version of mlua? 0.9 or 0.10? I find 0.10 to be way more stable with 0.9 having many deadlocks

**Reply by ElhamAryanpur** [reply]
> @ElhamAryanpur Which version of mlua? 0.9 or 0.10? I find 0.10 to be way more stable with 0.9 having many deadlocks

That is very true, I built it first at 0.9, then moved to 0.10 beta but ultimately the same result. So I went for the tokio handler gymnastics instead.

**Reply by khvzak** [reply]
@ElhamAryanpur do you have a full example please to reproduce? Would be helpful. In the code above the created Lua async function never called.

**Reply by ElhamAryanpur** [reply]
Oh yeah for sure, here I made a repository to show an example of something i was doing: https://github.com/ElhamAryanpur/mlua_async_potential_bug I didn't write the actix server code as by then it'd be a lot of work 😅

**Reply by ElhamAryanpur** [reply]
hope this helps @khvzak let me know if anything else is needed. Big respect for all the effort you've put on mlua

**Reply by khvzak** [reply]
@ElhamAryanpur few things
1) You don't need `std::thread::spawn` to make Lua VM non-blocking
2) No need to store tokio handle when you remove the thread
3) The `lua.load(include_str!("../program.lua")).exec()` must be `lua.load(include_str!("../program.lua")).exec_async().await` to execute registered async functions.

**Reply by ElhamAryanpur** [reply]
The thread was for my own use case as in my use I spawned cron jobs that executed lua code, which is where I can't share the lua VM to. Hence instead I spawned another thread and talked with cron jobs through channels instead.

But point 3, I didn't knew that existed! Thank you!

**Reply by ElhamAryanpur** [reply]
adding upon this, I tried the non thread method again to see if 0.10 have solved the issues, but seems not.

What I'm doing is like I said spawning tokio tasks for cron jobs that execute lua. Tried calling lua within again and the issue of lifetimes still exist, the VM need a lifetime longer that of the task sadly... Not even clone can solve it.

**Reply by khvzak** [reply]
There should be no issues in any version. 

This is an example how to run multiple tokio tasks concurrently in **v0.9**:
```rust
use mlua::{Lua, Result};
use std::rc::Rc;
use std::time::Duration;

async fn sleep(_: &Lua, ms: u64) -> Result<()> {
    tokio::time::sleep(Duration::from_millis(ms)).await;
    println!("Slept for {ms} ms");
    Ok(())
}

#[tokio::main]
async fn main() {
    let local = tokio::task::LocalSet::new();
    local
        .run_until(async {
            let lua = Rc::new(Lua::new());

            lua.globals()
                .set("sleep", lua.create_async_function(sleep).unwrap())
                .unwrap();

            // Run few tasks concurrently
            for i in 1..10 {
                let lua = lua.clone();
                tokio::task::spawn_local(async move {
                    lua.load(&format!("sleep({i})")).exec_async().await.unwrap();
                });
            }
        })
        .await;
    local.await;
}
```

This is example for **main** branch with `send` feature flag:
```rust
use mlua::{Lua, Result};
use std::time::Duration;

async fn sleep(_: Lua, ms: u64) -> Result<()> {
    tokio::time::sleep(Duration::from_millis(ms)).await;
    println!("Slept for {ms} ms");
    Ok(())
}

#[tokio::main]
async fn main() {
    let lua = Lua::new();

    lua.globals()
        .set("sleep", lua.create_async_function(sleep).unwrap())
        .unwrap();

    let mut tasks = tokio::task::JoinSet::new();
    for i in 1..10 {
        let lua = lua.clone();
        tasks.spawn(async move {
            lua.load(&format!("sleep({i})")).exec_async().await.unwrap();
        });
    }

    while let Some(_) = tasks.join_next().await {}
}
```

**Reply by cheesycod** [reply]
@khvzak Isn't lua single-threaded, so isnt the threadpool still needed due to the re-entrant mutex?

**Reply by khvzak** [reply]
@cheesycod Yes, Lua is single threaded and mlua holds an exclusive mutex to protect the vm. To run multiple blocking tasks in parallel you ideally need a pool of VMs. However you still should be able to concurrently execute futures even using a single threaded vm. The lock is obtained when `poll()` futures and released after.

**Reply by khvzak** [reply]
Also note that in non-`send` version there is no lock and the previous approach (with `LocalSet`) would work as well.

**Reply by ElhamAryanpur** [reply]
Interesting, I couldn't get that syntax to work at all with [tokio-cron](https://github.com/kurtbuilds/tokio-cron), perhaps the requirement for 'static in the signature was the issue:

```rs
impl Job {
    pub fn new<S, F, Fut>(cron: S, func: F) -> Self
        where
            F: Fn() -> Fut + Send + Sync + 'static,
            Fut: Future<Output=()> + Send + 'static,
            S: Into<String>
```

**Reply by ElhamAryanpur** [reply]
I think I'll keep the code I have with the thread for now, like you said ideally a pool of VM would be much better of an idea for parallel execution, so will try to get something like that up in the future. Except using tokio threads instead of std threads to get the async fixes per your suggestions
